std/algorithm/iteration.d

```diff
@@ -4280,3 +4284,218 @@ unittest
          [1, 2, 0],
          [2, 1, 0]]));
 }
+
+/**
+Finds the mean (colloquially known as the average) of a range. This function
+will return $(D real.nan) if the range is empty. This function is
+$(BIGOH r.length).
+
+Params:
+    r = an input range
+
+Returns:
+    the mean of r
+*/
+real mean(R)(R r)
+if (isInputRange!R &&
+    !isInfinite!R &&
+    is(ElementType!R : real))
+{
+    if (r.empty)
+    {
+        return real.nan;
+    }
+
+    static if (hasLength!R)
+    {
+        if (r.length == 1)
+        {
+            return r.front;
+        }
+
+        return sum(r) / real(r.length);
+    }
+    else
+    {
+        import std.typecons : tuple;
+
+        auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))
```

Comment:
If you got rid of the `static assert (isFloatingPoint!Result);` from `sumPairwise` you could do this:
```D
struct Pair(E)
{
    E e;
    size_t i;
    auto opBinary(string op : "+")(Pair!E rhs)
    {
        return Pair!E(e + rhs.e, i + rhs.i);
    }
}
auto sAndT = zip(repeat(1), r).map!(p => Pair!(ElementType!R)(p.expand)).sumPairwise();
return sAndT.e / sAndT.i;
```

But that's hardly ideal itself and to be honest I think it's OK as you have it already, as long as it's documented correctly.